### PR TITLE
Scalar Messaging: Expose join_rules and restrict to currently viewed room

### PR DESCRIPTION
**Pull request by lukebarnard1**
The post message API now:
- exposes a way of getting the join rules for a room (which must be public in some cases);
- returns an error if the ```room_id``` specified is not the one of the room currently being views.

(@Kegsay, I'm hoping you know how one could test this...)

### CODE REVIEW

**Comment by Kegsay** [comment]
ok to test

File: src/ScalarMessaging.js [hunk 1]
```diff
@@ -193,12 +194,16 @@ function getMembershipState(event, roomId, userId) {
     returnStateEvent(event, roomId, "m.room.member", userId);
 }
 
+function getJoinRules(event, roomId) {
+    console.log(`join_rules of ${roomId} requested.`);
+    returnStateEvent(event, roomId, "m.room.join_rules");
```
Comment (Kegsay) [inline src/ScalarMessaging.js]: Needs a `state_key` of `""`:
```javascript
returnStateEvent(event, roomId, "m.room.join_rules", "");
```

File: src/ScalarMessaging.js [hunk 2]
```diff
@@ -218,6 +223,18 @@ function returnStateEvent(event, roomId, eventType, stateKey) {
...
 
+var currentRoomId = null;
+
+// Listen for when a room is viewed
+dis.register(onAction);
+function onAction(payload) {
+    switch (payload.action) {
+        case 'view_room':
+            currentRoomId = payload.room_id;
+        break;
```
Comment (Kegsay) [inline src/ScalarMessaging.js]: I would just do:
```javascript
function onAction(payload) {
  if (payload.action !== "view_room") {
    return;
  }
  currentRoomId = payload.room_id;
}
```

We don't expect to add more listeners so the `switch` is a bit premature.

File: src/ScalarMessaging.js [hunk 3]
```diff
@@ -243,6 +260,15 @@ const onMessage = function(event) {
         sendError(event, "Missing room_id in request");
         return;
     }
+    if (!currentRoomId) {
+        sendError(event, "Must be viewing a room");
+        return;
+    }
+    if (roomId !== currentRoomId) {
+        sendError(event, "Room " + roomId + " not in view");
```
Comment (Kegsay) [inline src/ScalarMessaging.js]: s/not in view/not visible/

**Comment by Kegsay** [comment]
> (@Kegsay, I'm hoping you know how one could test this...)

Bwahahaha, no. Aside from running locally, there is little more you can do. The architecture makes mocking hard (singletons forces you to use something like `proxyquire`), and there's so many moving parts on the UI (when exactly does `view_room` fire?) that it's almost impossible to get good coverage here.

We *could* split this lot out into a class, inject the `MatrixClient` in the constructor (so we can mock it) and then inject `postMessage` requests, but I do not believe this file is complicated enough to warrant this (yet).

**Comment by Kegsay** [comment]
LGTM aside from `user_id` being required when it shouldn't be.

**Comment by lukebarnard1** [comment]
@Kegsay pls take another look, I had to add the set power level API and other stuff

File: src/ScalarMessaging.js [hunk 4]
```diff
@@ -188,17 +190,52 @@ function setBotOptions(event, roomId, userId) {
     });
 }
 
+function setBotPower(event, roomId, userId, level) {
+    if (!(Number.isInteger(level) && level >= 0)) {
+        sendError(event, "Power level must be positive integer.");
+        return;
+    }
+
+    console.log(`Received request to set power level for bot ${userId} in room ${roomId}.`);
```
Comment (Kegsay) [inline src/ScalarMessaging.js]: To what level?

**Comment by Kegsay** [comment]
LGTM aside from comment.

**Comment by Kegsay** [comment]
LGTM

**Pull request merged**